### PR TITLE
[WIP] Add noisy network and distributional Q-learning to implement Rainbow

### PR DESCRIPTION
**Pull request by joneswong**
## What do these changes do?

- [x] develop noisy network (implemented and in test now)
- [x] develop distributional Q-learning (implemented and in test not)

for DQN to enable Rainbow which is a combination of 6 improvements over DQN.

### CODE REVIEW

File: python/ray/rllib/agents/impala/impala.py [hunk 1]
```diff
@@ -37,7 +37,14 @@
 
     # Learning params.
     "grad_clip": 40.0,
-    "lr": 0.0001,
+    # either "adam" or "rmsprop"
```
Comment (ericl) [inline python/ray/rllib/agents/impala/impala.py]: Could you rebase this on master? It looks like these changes are already in.

**Comment by ericl** [comment]
@joneswong shall we publish some benchmarks on https://github.com/ray-project/rl-experiments for DQN vs Rainbow? I've started populating that repo with IMPALA and A2C benchmarks.

**Comment by ericl** [comment]
Also you should do a `git pull . upstream/master` to make the diff show up correctly.

**Comment by joneswong** [comment]
sorry for messy commits. I created a new pr from another clean branch, so this could be closed.

**Pull request closed without merging**